NEP 41: First steps towards improved Datatype Support (#15506)
numpy/numpy: doc/neps/nep-0041-improved-dtype-support.rst

these changes.
Since datatype changes touch a large part of code and behaviours, NEP 40
reviews some of the concepts, issues, and the current implementation.

Motivation and the Need for New User-Defined Datatypes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The current ecosystem has very few user defined datatypes using NumPy, the
two most promient being: ``rational`` and ``quaternion``.
These represent fairly simple datatypes which are not as strongly impacted
but the current limitations.
However, the number of currently available user defined datatypes
is in strong contrast with the need for datatypes such as:

* categorical types
* bfloat16, used in deep learning
* physical units (such as meters)
* extending e.g. integer dtypes to have a sentinel NA value
* geometrical objects [pygeos]_
* datatypes for tracing/automatic differentiation
* high or arbitrary precision math
* ….

Some of these are partially solved; for example unit capability is provided
in ``astropy.units``, ``unyt``, or ``pint``. However, these have to subclass
or wrap ``ndarray`` whereas special datatypes would provide a much more natural

> representation, and would immediately allow usage within tools such as

Reviewer: This would be a great place to add some example code to show the more natural representation.

Reviewer: Actually I think you may want two separate code examples: one aimed at users of a new dtype, and one aimed at authors of libraries that would want to create new dtypes.

Author: Well, I left those out intentionally. There are too many moving parts and things to bike shed, IMO.

What I want to convince you of here is:

1. We want "full featured dtypes"
2. We currently cannot have them
3. These are *necessary* decisions we can get out of the way before we do the detailed API design.

It may be that the text further down is too detailed in some cases, and thus hides that intention? Or do I need an example to show that `Dask` should simply work with a new `dtype`, but working with a new `array-like` such as `astropy.units` is probably less straight forward?

Author: Well, I tried to clarify the part of why I think subclasses are not a real solution a bit, maybe that helps. I am thinking of rather than adding a lot of new things, I may delete a bit of the discussions below (the API breakage is cleaned up a bit).

This means that in a first step the ``DType`` classes will be added, with
all, or most, new exposed API points giving a ``PreliminaryDTypeAPIWarning``.
This means starting the implementation further defined in NEP 42.

This allows for smaller patches and further future changes. In these first
steps no, or only very limited, new C-API shall be exposed.
The addition of new ``DTypes`` will then allow to address other changes
more incremental:

1. A new machinery for array coercion, with the goal of enabling user DTypes
   to behave in a full featured manner.
2. The replacement or wrapping of the current casting machinery.
3. Incremental redifinition of the current ``ArrFunctions`` slots into
   DType method slots.

Parallel to these, after step 1. is finished, the Phase II of revising the
UFunc machinery can be addressed.

In particular the step of creating a C defined ``DTypeMeta`` class with its
instances being ``DTypeClasses`` as mentioned above is a necessary first step
with useful longterm semantics.
This ``DTypeMeta`` must thus be implemented before being widely used to
restructure or enhance current code.

Backward compatibility
----------------------

> 

Reviewer: I would suggest to split this in impact for Python API users and C API users.

This will be written as ``Unit[float64]``, the unit itself is part of the
DType instance ``Unit[float64]("m")`` us a ``float64`` with meters attached::

    >>> meters = np.array([1, 2, 3], dtype=np.float64) * unit.m  # meters
    >>> print(meters)
    array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))

Note that units are a bit tricky, since it is debatable, whether::

    >>> np.array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))

should be valid syntax (coercing the float scalars without a unit to meters).
Once the array is created, math will work without any issue::

    >>> meters * 2 * unit.seconds
    array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))

Casting is not valid from one unit to the other, but can be between different
scales of the same dimensionality (although this may usually be "unsafe")::

    >>> meters.astype(Unit[float64]("s"))
    TypeError: Cannot cast meters to seconds.
    >>> meters.astype(Unit[float64]("km"))
    >>> meters.astype(meters.dtype.to_cgs())

The above notation may be somewhat clumsy, in some cases, and functions
could be used to convert things otherwise.
There may be ways to make these more convenient, but those must be left

> for future discussions::

Reviewer 2: I'm glad to see this come up, even if the eventual decision is not to support these method names.

One caution for deciding a set of method names is that it might make it difficult for existing unit libraries to migrate to using ufuncs, since they would either need to abandon their existing API, happen to have the exact same API you decide to use for NumPy, or decide to keep using an ndarray subclass or wrapper class.

One thing that at least at one point was in the realm of possibility was adding a way for ufuncs to define methods that then get attached to ndarray. I totally understand if the eventual decision is that it's too weird or surprising for a ufunc to be able to modify the ndarray API. However, that's just about the only way I can think of for it to be possible for all existing unit libraries to migrate to this functionality without introducing API breakages.

Author: Yes, it may be possible to make the quantity subclass nothing (or little) more than a MixIn.

For a dtype to modify the ndarray seems weird, but I have not figured out how weird I think it is (I understand pandas to uses the dtype to indicate the correct subclass). While I like the dtype to be super central, the tuple `(ndarray, DType)` defines the actual theoretical type (space of instances/values) that we are dealing with. My vision is that we want to approach a world where `(Sparse, DType)` is plausible. That implies `DType` must know nothing about `ndarray` explicitly and prefers unit to be a `DType`.
But: how `(ndarray, DType)` *looks* from python is a slightly different discussion, maybe!

From that standpoint, you could think that a `DType` provides a MixIn which is always used and induces(?)  a subclass of `ndarray`. (Theoretically `(ndarray, np.dtype[float64])` *is* a subtype, so why not a subclass.) Or, you could also have a `ndarray.elementwise.method` namespace that is dtype dependent.

I love disucussing it, but for now I think it should not be in here since it is a "next step", fully independent of the other thoughts.

=================================================
NEP 41 — First step towards a new Datatype System
=================================================

:title: Improved Datatype Support
:Author: Sebastian Berg
:Author: Add...
:Status: Draft
:Type: Standard Track
:Created: 2020-02-03

Abstract
--------

Datatypes in NumPy describe how to interpret each element in arrays.
For the most part, NumPy provides the usual numerical types,
as well as additional string and some datetime capabilities. 

> The growing Python community, however, has need for more more diverse datatypes.

Reviewer 3: `more more` -> `more` :)

cannot describe casting for such parametric datatypes implemented outside of NumPy.

NumPy currently has strings and datetimes as parametric datatypes.
These, more than the other datatypes, require special code paths in otherwise
generic code.
For user defined datatypes, this is not accessible, but also within NumPy it
means an increased complexity since the concerns of different datatypes
are not well separated.
This burden is exacerbated by the exposure of internal structures,
limiting even the development within NumPy such as the addition of new sorting methods.

Thus, there are many factors which limit the creation of new user defined
datatypes:

* Creating casting rules for parametric user types is either impossible
  or so complex that it has never been attempted.
* Type promotion, e.g. the operation deciding that adding float and integer
  values should return a float value, is very valuable for datatypes defined within NumPy,
  but is limited in scope for user defined and especially parametric datatypes.
* There is a general issue that much of the logic is written in single functions
  instead of being split as methods on the datatype itself.

> * Int the current design datatypes cannot have methods that do not generalize

Reviewer 4: ```suggestion
* In the current design datatypes cannot have methods that do not generalize
```

Similarly, Pandas has made a push into the same direction and undoubtedly
the community would be best served if such new features could be common
between NumPy, Pandas, and other projects.

Plan to Approach this Refactor
""""""""""""""""""""""""""""""

To address these issues in NumPy and enable new datatypes,
multiple development stages are required:

* Phase I: Restructure and extend the datatype infrastructure 

  * Organize Datatypes like normal Python classes

* Phase II: Incrementally define or rework API

  * Create a new and easily extensible API for defining new datatypes
    and related functionality.

  * Incrementally define all necessary functionality through the new API:

    * Defining operations such as ``np.common_type``.
    * Allowing to define casting between datatypes.
    * Add functionality necessary to create a numpy array from Python scalars
      (i.e. ``np.array(...)``).
    * …

  * Restructure how universal functions functions to:

    * make it possible to allow functions such as ``np.add`` to be extended by
      user defined datatypes such as Units.

    * allow efficient lookup for the correct implementation for user defined datatypes.

>     * enable reuse of existing code: E.g. units should be able to use the normal

Reviewer 3: `code: E.g. units` -> `code, e.g. units`

2. The current datatypes will be instances of these classes.
   All methods, which are currently defined on the instance, should instead be
   defined on the class. Storage information such as itemsize and byteorder
   are stored on the instance. Making a DType a class, allows for DType
   specific information to be stored more naturally.
3. The current NumPy scalars will *not* be instances of datatypes.
4. All new API provided to the user will hide implementation as much as
   possible. The public API should be identical but may be more limited,
   to the API used for the internal NumPy datatypes.
5. The UFunc machinery will be changed to replace the current dispatching
   and type resolution system (part of Phase II).
   The old system should be *mostly* supported as a legacy version for some time.
   This should thus not affect most users, but is a necessary large refactor.
   
Additionally, as a general design principle, the addition of new user defined
datatypes shall *not* change the behaviour of programs.
For example ``common_dtype(a, b)`` must not be ``c`` unless ``a`` or ``b`` know
that ``c`` exists.

Motivation and the Need for New User-Defined Datatypes

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Author: Thanks! Maybe thats why github does not render it, pandoc did not warn me about it, I guess different dialects.

The following operations may be desirable but are *not* straight forward::

    >>> table2 = np.array([Breakfast.spam, Breakfast.eggs])  # discover same dtype
    >>> table == Breakfast.spam
    array([True, False, False])

To define these NumPy would need to find the correct Enum DType.
And it is unclear how that should be done, because the type associated with
an ``EnumDType`` could be any ``Enum`` subclass.

An alternative approach, which solves those issues, is to create the DTypes as::

    >>> BreakfastDType = EnumDType(Breakfast)  # class/DType factory
    >>> issubclass(BreakfastDType, EnumDType)
    True

Where ``BreakfastDType`` is a subclass and not just an instance of ``EnumDType``.
Which makes sense since also ``Enum`` is a class factory.
The Enum example is particularly complicated (it may be good to keep in mind,
but it is likely that is would not be supported initially).

Unit on the Datatype
""""""""""""""""""""

There are different ways to define Units, depending on how the internal
machinery would be organized, one way is to have a single Unit datatype
for every existing numerical type.
This will be written as ``Unit[float64]``, the unit itself is part of the

> DType instance ``Unit[float64]("m")`` us a ``float64`` with meters attached::

Author: There is a space of possibilities here, note that the above notation: `Unit[float64]` denotes a  `UnitFloat64` class. You could allow `Unit(float64, "m")` to return such a class as well of course. (A class factory returning subclasses)

I am currently operating under the mode that you would define the (C-level) math on the `Unit[float64]`, so it is the natural "class" (provider of methods, and ufuncs).
It is likely that in the end there will be no clear distinction and you can do what you want (although the above would probably still be the preferred approach).

    >>> meters = np.array([1, 2, 3], dtype=np.float64) * unit.m  # meters
    >>> print(meters)
    array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))

Note that units are a bit tricky, since it is debatable, whether::

    >>> np.array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))

should be valid syntax (coercing the float scalars without a unit to meters).
Once the array is created, math will work without any issue::

    >>> meters * 2 * unit.seconds

>     array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))

Reviewer 2: FWIW, unyt handles all of these aliases by canonicalizing unit names in the parser. I agree that `ms` should be milliseconds...

>     array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))

Author: ```suggestion
    array([2.0, 4.0, 6.0], dtype=Unit[float64]("m s"))
```

>     array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))

Reviewer 5: Indeed, interaction between the various unit packages seems to be best done via their string representation... But besides the point here: just "m s" is clear enough here.

Casting is not valid from one unit to the other, but can be between different
scales of the same dimensionality (although this may usually be "unsafe")::

    >>> meters.astype(Unit[float64]("s"))
    TypeError: Cannot cast meters to seconds.

>     >>> meters.astype(Unit[float64]("km"))

Reviewer 6: On second read, it's not very clear to me how this would work under the hood.
Does it imply that the Unit class will need to define something like a magic method
```python
def __astype__(self, array: np.ndarray, newtype: numpy.DType) -> np.ndarray:
```
?
What happens if the two types belong to different libraries? Would it be possible to define
```python
def __fromtype__(self, array: np.ndarray) -> np.ndarray:
```
in the target dtype if the ``__astype__`` before in library A (which defined array.dtype) raises TypeError?

Author: They would provide `CastingImpl` objects (or maybe rather C-structs with a specific, extensible initialization syntax) which houses all the necessary information, mainly a way to check casting safety and "adjust" the dtypes and then an actual loop doing the cast. These things are in NEP 42, although that will needs a bit settling on the API and revising on the text, I think the general notions are sound by now.

That can be extended later (similar to python which has `__getattr__` with which you can get an object even though it is not part of the class dictionary), but initially I want you to provide a list of possible casts at creation time: I.e. one of the DTypes must know about the other one.

I think that is a good "basic" design choice, because I do not really want users to discourage messing with the casting of datatypes you did not define.

>     >>> meters.astype(meters.dtype.to_cgs())

Reviewer 2: centimeter-gram-second

>     >>> meters.astype(meters.dtype.to_cgs())

Reviewer 7: Possibly worth expanding in a comment or sentence before this example.

    dtype), being valid.
    At least in some code paths, a similar mechanism is already used.

  * The ``scalarkind`` slot and registration of scalar casting will be
     removed/ignored without replacement.
     It currently allows partial value based casting.
     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
     but will not be used internally and be deprecated.

   * Current user dtypes are specifically defined as instances of ``np.dtype``.
     The creation works by the user providing a prototype instance.
     NumPy will need to modify at least the type during registration.
     This has no effect for either ``rational`` or ``quaternion`` and mutation
     of the structure seems unlikely after registration.

Since there is a fairly large API surface concerning datatypes, further limitations
or the limitation of a certain function to currently existing datatypes is
likely to occur.
For example current functions which currently use the type number as input
should probably be replaced with functions taking DType classes instead
in the long term.
Although public, large parts of this C-API seems very rarely and possibly

> completely used by downstream projects.

Reviewer 6: ```suggestion
never used by downstream projects.
```

=================================================
NEP 41 — First step towards a new Datatype System
=================================================

:title: Improved Datatype Support
:Author: Sebastian Berg
:Author: Add...
:Status: Draft
:Type: Standard Track
:Created: 2020-02-03

.. note::

    This NEP is part of a series of NEPs encompassing first information
    about the previous dtype implementation and issues with it in NEP 40.

>     NEP 41 (this document) then provides an overview and generic design

Reviewer 7: Can we make these links?

Author: Yes, but only as they are being merged, and I hope to merge this one first.

    choices for the refactor.  
    Further NEPs 42 and 43 go into the technical dtypes of the datatype
    and universal function related internal and external API changes.
    In some cases it may be necessary to consult the other NEPs for a full
    picture of the desired changes and why these changes are necessary.

Abstract
--------

`Datatypes <data-type-objects-dtype>` in NumPy describe how to interpret each
element in arrays. NumPy provides ``int``, ``float``, and ``complex`` numerical

> types, as well as string, datetime, and structured (or `~numpy.recarray`)

Reviewer 7: I don't think we need to mention recarray here. Can we just link to the docs for structured dtypes, which work just fine without recarray?

datatype capabilities.  The growing Python community, however, has need for
more diverse datatypes.
Examples are datatypes with unit information attached (such as meters) or
categorical datatypes (fixed set of possible values).
However, the current NumPy datatype API is too limited to allow the creation
of these.

This NEP is the first step to enable such growth; it will lead to
a simpler development path for new datatypes. Support is also proposed to
create datatypes from Python rather than in C.

> By refactoring our datatype API and improving its maintainability,

Author: Thanks, tried it slightly different, do not want to dangle new numpy datatypes though.

future development will become possible, not only for external user datatypes,
but also within NumPy.

Motivation and Scope
--------------------

See also the user impact section for examples of what kind of new datatypes
will be enabled by the proposed changes in the long run.

Motivation
^^^^^^^^^^

One of the main issues with the current API is the definition of typical
functions such as addition and multiplication for parametric datatypes
(compare also NEP 40) require additional steps to determine the output type.

> For example when adding two strings of length four, the result is a string

Reviewer 7: ```suggestion
For example when adding two strings of length 4, the result is a string
``for consistency with `8` below

of length 8, which is different from the input.
Similarly, a datatype which embeds a physical unit must calculate the new unit
information: dividing a distance by a time results in a speed.
A related difficulty is that the current casting rules

> – the conversion between different datatypes –

Reviewer 7: ```suggestion
---the conversion between different datatypes---
```
Possibly best to avoid Unicode and let Sphinx handle the mdash.

Author: Changing to `--` as en dash with spaces for now,  if unicode might make problems. I personally prefer that spelling (I know em dash without spaces is more common in current english).

cannot describe casting for such parametric datatypes implemented outside of NumPy.

For user defined datatypes, these extra steps are not accessible, but also
within NumPy they require increased complexity.

> In general the concerns of different datatypes are not well separated.

Reviewer 7: Encapsulated sounds good here, after about hyphenation.

This burden is exacerbated by the exposure of internal C structures,

> limiting for example the addition of new sorting methods.

Reviewer 7: ```suggestion
limiting the addition of new fields (for example new sorting methods).
```

The large need to solve these issues has driven the scientific community
to create work-arounds in multiple projects implementing physical units as an
array-like class instead of a datatype, which would generalize better across
multiple array-likes (Dask, pandas, etc.).
Already, Pandas has made a push into the same direction with its
extension arrays [pandas_extension_arrays]_ and undoubtedly
the community would be best served if such new features could be common
between NumPy, Pandas, and other projects.

Scope
^^^^^

This NEP proposes to move ahead with the necessary creation of new dtype
subclasses (see Phase I Implementation section),
and start working on implementing current functionality within the new context
Within the context of this NEP all development will be fully private API or
use preliminary underscored names which must be changed in the future.
Most of the internal and public API choices are part of a second Phase
and will be discussed in more detail in the following NEPs 42 and 43.
The initial implementation of this NEP will have little or no effect on users,
but provides the necessary ground work for incrementally addressing the
full rework.

> The implementation of this NEP and leading up to a large rework of how

Author: Changed it to:
```
The implementation of this NEP and the following, implied large rework of how
datatypes are defined in NumPy is expected to create small incompatibilities
```

datatypes are defined in NumPy is expected to create small incompatibilities
(see backward compatibility section).
However, a transition requiring large code adaption is not anticipated and not
within scope.

Specifically, this NEP makes the following design choices which are discussed
in more details in the detailed description section:

1. Each datatype will be a class instance of a DType class, with most of the
   datatype-specific logic being implemented
   as special methods on the class. In the C-API, these correspond to specific

>    slots. In short ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``.

Reviewer 7: ```suggestion
   slots. In short, for ``f = np.dtype("f8")``, ``isinstance(f, np.dtype)`` will remain true, but ``type(f)`` will be a subclass of ``np.dtype`` rather than just ``np.dtype`` itself.
```

   All the methods which are currently stored on the instance, should instead

>    be defined on the class as is normal in Python.

Reviewer 7: Today they're already defined on the class though, right? Can you name a couple of the functions as example here?

Author: Yeah, I am talking about ``PyArray_ArrFuncs`` slots, which of course are, in a sense, defined on something like a class, except that there is no class object itself.

   Storage information such as itemsize and byteorder can differ between
   different dtype instances (e.g. "S3" vs. "S8") and will remain part of the instance.
   This means that in the long run the current lowlevel access to dtype methods

>    will be removed (see ``ArrFunctions`` in NEP 40).

Reviewer 7: ```suggestion
   will be removed (see ``PyArray_ArrFuncs`` in NEP 40).
```

2. The current NumPy scalars will *not* change, they will not be instances of
   datatypes. This will also be true for new datatypes, scalars will not be

>    instances of a dtype (although ``isinstance(scalar, dtype)`` may be defined).

Reviewer 7: By "be defined", do you mean "made to return `True` when appropriate"?

With detailed technical decisions being followed up in NEP 42.

Further, the public API will be designed in a way that is extensible in the future:

3. All new API functions provided to the user will hide implementation details
   as much as possible. The public API should be an identical, but limited,

>    version of the API used for the internal NumPy datatypes.

Reviewer 7: Can we be clear about C vs Python API here?

Author: Right added the `C-`, Python is not very relevant for now.

The changes to the datatype system in Phase II must include a large refactor of the
UFunc machinery, which will be further defined in NEP 43:

4. To enable all of the desired functionality for new user defined datatypes,
   the UFunc machinery will be changed to replace the current dispatching
   and type resolution system.
   The old system should be *mostly* supported as a legacy version for some time.

Additionally, as a general design principle, the addition of new user defined
datatypes will *not* change the behaviour of programs.
For example ``common_dtype(a, b)`` must not be ``c`` unless ``a`` or ``b`` know
that ``c`` exists.

User Impact
-----------

The current ecosystem has very few user defined datatypes using NumPy, the
two most prominent being: ``rational`` and ``quaternion``.
These represent fairly simple datatypes which are not strongly impacted
by the current limitations.
However, we have identified a need for datatypes such as:

* bfloat16, used in deep learning
* categorical types
* physical units (such as meters)
* datatypes for tracing/automatic differentiation
* high, fixed precision math

> * specifialized integer types such as int2, int24

Reviewer 7: ```suggestion
* specialized integer types such as int2, int24
```

* new, better datetime representations
* extending e.g. integer dtypes to have a sentinel NA value
* geometrical objects [pygeos]_

Some of these are partially solved; for example unit capability is provided
in ``astropy.units``, ``unyt``, or ``pint``, as `numpy.ndarray` subclasses.
Most of these datatypes, however, simply cannot be reasonably defined
right now.
An advantage of having such datatypes in NumPy is that they should integrate
seamlessly with other array or array-like packages such as Pandas,
``xarray``, or ``Dask``.

The long term user impact of implementing this NEP will be to allow both
the growth of the whole ecosystem by having such new datatypes, as well as
consolidating implementation of such datatypes within NumPy to achieve
a better interoperability.

Examples

> ^^^^^^^^

Reviewer 7: Perhaps worth a disclaimer saying that the exact behavior of these examples (eg the choices of casting rules) are not in the scope of this NEP - merely they're an example of the type of choice we want to allow users to be able to make.

Simple Numerical Types
""""""""""""""""""""""

Mainly used where memory is a consideration, lower-precision numeric types

> such as ``bfloat16`` are common in other computational frameworks.

Reviewer 7: A reference link to `bfloat16` would be handy here.

For these types the definitions of things such as ``np.common_type`` and
``np.can_cast`` are some of the most important interfaces. Once they
support ``np.common_type``, it is (for the most part) possible to find
the correct ufunc loop to call, since most ufuncs – such as add – effectively
only require ``np.result_type``::

    >>> np.add(arr1, arr2).dtype == np.result_type(arr1, arr2)

> and ``np.result_type`` is largely identical to ``np.common_type``.

Reviewer 7: Do we want something like:

```suggestion
and `~numpy.result_type` is largely identical to `~numpy.common_type`.
```

Fixed, high precision math
""""""""""""""""""""""""""

Allowing arbitrary precision or higher precision math is important in
simulations. For instance ``mpmath`` defines a precision::

    >>> import mpmath as mp
    >>> print(mp.dps)  # the current (default) precision
    15

NumPy should be able to construct a native, memory efficient array from
a list of ``mpmath.mpf`` floating point objects::

    >>> arr1 = np.array(mp.arange(3))  # (mp.arange returns a list)
    >>> print(arr)  # Must find the correct precision from the objects:

>     array(['0.0', '1.0', '2.0'], dtype=mpf[dps=15])

Reviewer 7: ```suggestion
    >>> arr_15_dps = np.array(mp.arange(3))  # (mp.arange returns a list)
    >>> print(arr_15_dps)  # Must find the correct precision from the objects:
    array(['0.0', '1.0', '2.0'], dtype=mpf[dps=15])
```

We should also be able to specify the desired precision when
creating the datatype for an array. Here, we use ``np.dtype[mp.mpf]``
to find the DType class (the notation is not part of this NEP),
which is then instantiated with the desired parameter.
This could also be written as ``MpfDType`` class::

    >>> arr2 = np.array([1, 2, 3], dtype=np.dtype[mp.mpf](dps=100))
    >>> print(arr1 + arr2)

>     array(['0.0', '2.0', '4.0'], dtype=mpf[dps=100])

Reviewer 7: ```suggestion
    >>> arr_100_dps = np.array([1, 2, 3], dtype=np.dtype[mp.mpf](dps=100))
    >>> print(arr_15_dps + arr_100_dps)
    array(['0.0', '2.0', '4.0'], dtype=mpf[dps=100])
```

The ``mpf`` datatype can decide that the result of the operation should be the
higher precision one of the two, so uses a precision of 100.
Furthermore, we should be able to define casting, for example as in::

    >>> np.can_cast(arr1.dtype, arr2.dtype, casting="safe")
    True
    >>> np.can_cast(arr2.dtype, arr1.dtype, casting="safe")
    False
    >>> np.can_cast(arr2.dtype, arr2.dtype, casting="same_kind")

>     True

Reviewer 7: Above changes are to make this easier to follow
```suggestion
    >>> np.can_cast(arr_15_dps.dtype, arr_100_dps.dtype, casting="safe")
    True
    >>> np.can_cast(arr_100_dps.dtype, arr_15_dps.dtype, casting="safe")
    False  # loses precision
    >>> np.can_cast(arr_100_dps.dtype, arr_100_dps.dtype, casting="same_kind")
    True
```

    >>> np.multiply(meters, seconds).dtype != np.result_type(meters, seconds)

In fact ``np.result_type(meters, seconds)`` must error without context
of the operation being done.
This example highlights how the specific ufunc loop
(loop with known, specific DTypes as inputs), has to be able to to make
certain decisions before the actual calculation can start.

Implementation
--------------

Plan to Approach the Full Refactor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To address these issues in NumPy and enable new datatypes,
multiple development stages are required:

* Phase I: Restructure and extend the datatype infrastructure (This NEP)

  * Organize Datatypes like normal Python classes [`PR 15508`]_

* Phase II: Incrementally define or rework API

  * Create a new and easily extensible API for defining new datatypes
    and related functionality. (NEP 42)

  * Incrementally define all necessary functionality through the new API (NEP 42):

    * Defining operations such as ``np.common_type``.
    * Allowing to define casting between datatypes.
    * Add functionality necessary to create a numpy array from Python scalars
      (i.e. ``np.array(...)``).
    * …

>   * Restructure how universal functions functions to (NEP 43):

Reviewer 7: Perhaps you meant
```suggestion
  * Restructure how universal functions function (NEP 43), to:
```

      extended by user defined datatypes such as Units.

    * allow efficient lookup for the correct implementation for user defined
      datatypes.

    * enable reuse of existing code. Units should be able to use the
      normal math loops and add additional logic to determine output type.

* Phase III: Growth of NumPy and Scientific Python Ecosystem capabilities:

  * Cleanup of legacy behaviour where it is considered buggy or undesirable.
  * Provide a path to define new datatypes from Python.
  * Assist the community in creating types such as Units or Categoricals
  * Allow strings to be used in functions such as ``np.equal`` or ``np.add``.
  * Remove legacy code paths within NumPy to improve long term maintainability

This document serves as a basis for phase I and provides the vision and
motivation for the full project.
While of limited use to the end user, the first step involves a conceptional
cleanup of the current datatype system. It provides a more "pythonic" datatype
Python type object, with a clear class hierarchy.

The second phase is the incremental creation of all APIs necessary to define
fully featured datatypes and reorganization of the NumPy datatype system.

> Phase I concerns mainly private APIs, while phase II is expected to include

Reviewer 8: The wording here might invite questions about this development plan - might want to be less specific and just mention that phase 2 is concerned with the design and development of the public API.

a slow creation of a *preliminary* public API which will develop into the
final stable API.

Some of the benefits of a large refactor may only become evident after the full
deprecation of the current legacy implementation (i.e. larger code removals).
However, these steps are necessary for improvements to many parts of the
core NumPy API, and are expected to make the implementation generally
easier to understand.

> Pushing forward with Phase I cleans up the concepts of the API to coincide

Reviewer 8: > Pushing forward with Phase I cleans up the concepts of the API to coincide
> with how Python is typically organized and unlocks the ability to improve
> parts of NumPy which currently are not extensible.

This whole sentence is a bit redundant and in my opinion could be safely removed

with how Python is typically organized and unlocks the ability to improve
parts of NumPy which currently are not extensible.

> The following picture gives another overview of the overall design idea.

Reviewer 8: Wording suggestion for these sentences, something along the lines of:

The following figure illustrates the proposed designed at a high level, and roughly delineates the components of the overall design expressly addressed in this NEP.

Author: Reused your sentence, removed the last bit though

Note that this NEP only regards Phase I (shaded area),
the rest encompasses Phase II and the design choices are up for discussion,
however, it highlights that the DType datatype class is the central, necessary
concept:

.. image:: _static/nep-0041-mindmap.svg

First steps directly related to this NEP
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The required changes necessary to NumPy are massive and touch a large part
of the code base.

> We thus propose the implementation of the above steps listed in Phases I and II.

Reviewer 8: This sentence is a bit unclear. The point that I got from these two sentences was the fact that the changes will be so extensive is what persuaded you to take a multi-stage approach.

Although it is possible that new DTypes will only be useful after Phase II
is finished, this NEP proposes to start implementation in an incremental way.
This means that in a first step the ``DType`` classes will be added, with

> all, or most, new exposed API points giving a ``PreliminaryDTypeAPIWarning``.

Reviewer 8: Is it expected that *users* of NumPy (i.e. not doing anything with the C-API) would ever see these warnings?

This allows for smaller patches and further future changes. In these first
steps, only a very limited new C-API will be exposed to allow writing tests,

> with no expectation that early adopters will begin using it.

Reviewer 8: Is it just that you don't expect anyone to use the new, limited API, or that you actively want to *discourage* people from using the limited API until phase I is closer to complete?

The addition of ``DType`` will then enable addressing other changes
more incrementally:

1. New machinery for array coercion, with the goal of enabling user DTypes
   with appropriate class methods.
2. The replacement or wrapping of the current casting machinery.
3. Incremental redefinition of the current ``ArrFunctions`` slots into
   DType method slots.

> Parallel to these, after step 1. is finished, Phase II of revising the

Reviewer 8: The mixing of "steps" and "phases" here could be confusing. Consider replacing "step 1" with "new machinery for array coercion is in place"

UFunc machinery can be started.

> In particular the step of creating a C defined ``DTypeMeta`` class with its

Reviewer 8: C-defined

Author: I have restructured that section a bit, maybe you can have a look later (after I pushed).

instances being ``DTypeClasses`` is a necessary first step to add the ability
of storing custom slots on the datatype in C.
This ``DTypeMeta`` must thus be implemented before being widely used to
restructure or enhance current code, thus we propose to proceed with mainly
private additions to the DType classes.

Backward compatibility
----------------------

While the actual backward compatibility impact is not yet fully clear,

> we anticipate, and accept the following changes:

Reviewer 8: Are these backward-compatibility considerations related only to Phase I (i.e. NEP 41-related) changes? Or is this an accounting of changes that are expected to result from the entire refactor? If the former, you should explicitly state that.

Author: Made it explicitly Phase I and II, since I personally think it is not useful to only list Phase I impact.

* **Python API**:

  * ``type(np.dtype("f8"))`` will be a subclass of ``np.dtype``, while right
    now ``type(np.dtype("f8")) is np.dtype``.
    Code should use ``isinstance`` checks, and in very rare cases may have to
    be adapted to use it.

* **C-API**:

    * In old versions of NumPy ``PyArray_DescrCheck`` is a macro which uses
      ``type(dtype) is np.dtype``. When compiling against an old NumPy version,
      the macro may have to be replaced with the corresponding
      ``PyObject_IsInstance`` call. (If this is a problem, we could backport
      fixing the macro)

   * The UFunc machinery changes will break *limited* parts of the current
     implementation. Replacing e.g. the default ``TypeResolver`` is expected
     to remain supported for a time, although optimized masked inner loop iteration

>      (which is not even used *within* NumPy) is expected to not remain supported

Reviewer 8: "not remain supported" -> "lose support" or "will no longer be supported"

     and lead to errors instead.

>    * All current functions defined on the dtypes currently, such as

Reviewer 8: double-usage of "current/currently"

>      ``PyArray_Descr->f->nonzero``, will move to a new home and recieve new

Reviewer 8: Is there a more precise term for "will move to a new home"? Perhaps "will be defined elsewhere"?

"and receive new ways to access them" -> "will be accessed differently"?

     ways to access them.

>      This means that in the long run some (rare) lowlevel access code will

Reviewer 8: Consider removing "(rare)" and adding a new sentence at the end: "Such changes are expected to be necessary very infrequently"

     have to be changed to use the new API.

* **dtype implementors (C-API)**:

  * The array that is currently provided to some functions (such as cast functions),

>     may not be provided anymore generally.

Reviewer 8: "may not be provided anymore generally" -> "may no longer be provided"

    For example ``PyArray_Descr->f->nonzero`` or ``PyArray_Descr->f->copyswapn``,
    may instead receive a dummy array object with only some fields (mainly the
    dtype), being valid.
    At least in some code paths, a similar mechanism is already used.

  * The ``scalarkind`` slot and registration of scalar casting will be
     removed/ignored without replacement.

>      It currently allows partial value based casting.

Reviewer 8: value-based

     The ``PyArray_ScalarKind`` function will continue to work for builtin types,
     but will not be used internally and be deprecated.

   * Currently user dtypes are defined as instances of ``np.dtype``.
     The creation works by the user providing a prototype instance.
     NumPy will need to modify at least the type during registration.
     This has no effect for either ``rational`` or ``quaternion`` and mutation
     of the structure seems unlikely after registration.

Since there is a fairly large API surface concerning datatypes, further changes

> or the limitation of a certain function to currently existing datatypes is

Reviewer 8: "or the limitation of a certain function" -> "or limiting certain functions"

likely to occur.
For example functions which use the type number as input
should be replaced with functions taking DType classes instead.

> Although public, large parts of this C-API seems very rarely and possibly

Reviewer 8: "seems very rarely and possibly never used by downstream projects" -> "seem to be used rarely, possibly never, by downstream projects"

never used by downstream projects.

Detailed Description

> --------------------

Reviewer 8: It took me a couple readings to figure out that the "Detailed Description" subheadings correlated to the "design decisions" made in the "Scope" section. Consider explicitly stating so here in the introduction to the section, or at least linking the subheadings below back to the original listing of the design decisions.

> This section details some of the design decisions above and gives

Reviewer 8: Are these design decisions for Phase I or more general?

more details.

> Since datatype changes touch a large part of code and behaviours, NEP 40

Reviewer 8: > Since datatype changes touch a large part of code and behaviours

This is redundant given the previous section.

reviews some of the concepts, issues, and the current implementation.

Datatypes as Python Classes (1)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The current NumPy datatypes are not full scale python classes.

Reviewer 8: full-scale

They are instead (prototype) instances of a single ``np.dtype`` class.
Changing this means that any special handling, e.g. for ``datetime``
can be moved to the Datetime DType class instead, away from monolithic general
code (e.g. current ``PyArray_AdjustFlexibleDType``).

> The main API side effect of this is that special methods move

Reviewer 8: The main consequence of this change with respect to the API is that special methods move from the dtype instances to methods on the new dtype class.

from the dtype instances to class methods on the class.
This is the typical design pattern used in Python.

> Adding a new, pythonic point to store these methods and information, will

Reviewer 8: Organizing these methods and information in a more Pythonic way provides a solid foundation for refining and extending the API in the future.

allow us to further refine the API and allow it to grow in the future.
The current API cannot be extended due to how it is exposed publically.
This means for example that the methods currently stored in ``ArrFunctions``
on each datatype (see NEP 40) will be defined differently in the future and
deprecated in the long run.

The most prominent visible side effect of this will be that
``type(np.dtype(np.float64))`` will not be ``np.dtype`` anymore.
Instead it will be a subclass of ``np.dtype`` meaning that
``isinstance(np.dtype(np.float64), np.dtype)`` will remain true.
This will also add the ability to use ``isinstance(dtype, np.dtype[float64])``
thus removing the need to use ``dtype.kind``, ``dtype.char``, or ``dtype.type``
to do this check.

> If DTypes were full scale Python classes, the question of subclassing arises.

Reviewer 8: With the design decision of DTypes as full-scale Python classes,

Inheritance, however, appears problematic and a complexity best avoided
(at least initially) for container datatypes.

> Further, subclasses may be more interesting for the interoperatbility for

Reviewer 8: "the interoperability" -> "interoperability"

example with GPU backends (CuPy) storing additional methods related to the
GPU rather than as a mechanism to define new datatypes.

> Since a class hierarchy and subclass order provides value, phase I will allow

Reviewer 8: "subclass order" -> "method resolution order"?

Author: Removed mro does not make sense here

the creation of *abstract* datatypes.
An example for an abstract datatype would be the datatype equivalent of
``np.floating``, representing any floating point number.
These can serve the same purpose as Python's abstract base classes.

Scalars should not be instances of the datatypes (2)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For simple datatypes such as ``float64`` (see also below), it seems
tempting that the instance of a ``np.dtype("float64")`` can be the scalar.
This idea may be even more appealing due to the fact that scalars,
rather than datatypes, currently define a useful type hierarchy.

However, we have specifically decided against this for a number of reasons.

> First, the above described new datatypes would be instances of DType

Reviewer 8: "the above described new datatypes" -> "the new datatypes described herein"

classes.

> Making these instances themselves classes, while possible, adds an additional

Reviewer 8: "an additional complexity" -> "additional complexity"

complexity that users need to understand.
It would also mean that scalars must have storage information (such as byteorder)
which is generally unnecessary and currently is not used.
Second, while the simple NumPy scalars such as ``float64`` may be such instances,
it should be possible to create datatypes for Python objects without enforcing
NumPy as a dependency.
However, Python objects that do not depend on NumPy cannot be instances of a NumPy DType.
Third, there is a mismatch between the methods and attributes which are useful
for scalars and datatypes. For instance ``to_float()`` makes sense for a scalar
but not for a datatype and ``newbyteorder`` is not useful on a scalar (or has
a different meaning).

Overall, it seem rather than reducing the complexity, i.e. by merging

> the two distinct type hierarchies, making scalars instances of DTypes would

Reviewer 8: "would increase the complexity of both the design and implementation"

add complexity for the user and make the implementation more complex.

A possible future path may be to instead simplify the current NumPy scalars to

> be much simpler objects which largely derived their behaviour from the datatypes.

Reviewer 8: "derived" -> "derive"

C-API for creating new Datatypes (3)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The current C-API with which users can create new datatypes
is limited in scope, and requires use of "private" structures. This means
the API is not extensible: no new members can be added to the structure
without losing binary compatibility.
This has already limited the inclusion of new sorting methods into
NumPy [new_sort]_.

> The new version shall thus replace the current ``ArrFuncs`` structure used

Reviewer 7: ```suggestion
The new version shall thus replace the current ``PyArray_ArrFuncs`` structure used
```

For a more detailed accounting of the various phases, see
"Plan to Approach the Full Refactor" in the Implementation section below.
This NEP proposes to move ahead with the necessary creation of new dtype
subclasses (Phase I),
and start working on implementing current functionality.
Within the context of this NEP all development will be fully private API or
use preliminary underscored names which must be changed in the future.
Most of the internal and public API choices are part of a second Phase
and will be discussed in more detail in the following NEPs 42 and 43.
The initial implementation of this NEP will have little or no effect on users,
but provides the necessary ground work for incrementally addressing the
full rework.

The implementation of this NEP and the following, implied large rework of how
datatypes are defined in NumPy is expected to create small incompatibilities
(see backward compatibility section).
However, a transition requiring large code adaption is not anticipated and not
within scope.

Specifically, this NEP makes the following design choices which are discussed
in more details in the detailed description section:

> 1. Each datatype will be a class instance of a DType class, with most of the

Reviewer 7: Are you trying to say:
```suggestion
1. Each datatype will be an instance of a subclass of ``np.dtype``, with most of the
```
or is this phrasing deliberately different.

Author: No, that is much better to understand, thanks.

   datatype-specific logic being implemented
   as special methods on the class. In the C-API, these correspond to specific
   slots. In short, for ``f = np.dtype("f8")``, ``isinstance(f, np.dtype)`` will remain true, but ``type(f)`` will be a subclass of ``np.dtype`` rather than just ``np.dtype`` itself.

>    The ``PyArray_ArrFuncs`` which are currently stored as a pointer on the instance,

Reviewer 7: ```suggestion
   The ``PyArray_ArrFuncs`` which are currently stored as a pointer on the instance (as ``PyArray_Descr->f``),
```

Author: @eric-wieser would you be happy to be added as an author at this point?

Note that this NEP only regards Phase I (shaded area),
the rest encompasses Phase II and the design choices are up for discussion,
however, it highlights that the DType datatype class is the central, necessary
concept:

.. image:: _static/nep-0041-mindmap.svg

First steps directly related to this NEP
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The required changes necessary to NumPy are large and touch many areas
of the code base
but many of these changes can be addressed incrementally.

To enable an incremental approach we will start by creating a C defined
``PyArray_DTypeMeta`` class with its instances being the ``DType`` classes,
subclasses of ``np.dtype``.
This is necessary to add the ability of storing custom slots on the DType in C.
This ``DTypeMeta`` will be implemented first to then enable incremental
restructuring of current code.

The addition of ``DType`` will then enable addressing other changes
incrementally, some of which may begin before the settling the full internal
API:

1. New machinery for array coercion, with the goal of enabling user DTypes
   with appropriate class methods.
2. The replacement or wrapping of the current casting machinery.

> 3. Incremental redefinition of the current ``ArrFunctions`` slots into

Reviewer 7: ```suggestion
3. Incremental redefinition of the current ``PyArray_ArrFuncs`` slots into
```

should be replaced with functions taking DType classes instead.
Although public, large parts of this C-API seem to be used rarely,
possibly never, by downstream projects.

Detailed Description
--------------------

This section details the design decisions covered by this NEP.
The subsections correspond to the list of design choices presented
in the Scope section.

Datatypes as Python Classes (1)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The current NumPy datatypes are not full scale python classes.
They are instead (prototype) instances of a single ``np.dtype`` class.
Changing this means that any special handling, e.g. for ``datetime``
can be moved to the Datetime DType class instead, away from monolithic general
code (e.g. current ``PyArray_AdjustFlexibleDType``).

The main consequence of this change with respect to the API is that
special methods move from the dtype instances to methods on the new DType class.
This is the typical design pattern used in Python.
Organizing these methods and information in a more Pythonic way provides a
solid foundation for refining and extending the API in the future.
The current API cannot be extended due to how it is exposed publically.

> This means for example that the methods currently stored in ``ArrFunctions``

Reviewer 7: ```suggestion
This means for example that the methods currently stored in ``PyArray_ArrFuncs``
```

on each datatype (see NEP 40) will be defined differently in the future and
deprecated in the long run.

The most prominent visible side effect of this will be that

> ``type(np.dtype(np.float64))`` will not be ``np.dtype`` anymore.

Reviewer 7: Will it be the case that `type(np.dtype(np.float64)) is type(np.dtype(np.float32))`?

Or phrased another way, is this NEP deliberately leaving that question open?

Author: My intent is that is False, I am happy to define that right now. If it was not False, `type(np.dtype(np.float64))` would be parametric and I think it is desirable to simplify life by having non-parametric DTypes were possible. They seems easier to me to reason about, and to define ufunc dispatching for.
For the current buildin dtypes, it also seems only right, since they already have ArrFuncs and typenumbers that are distinct.

Of course in general it should be possible to do the above, although I expect some methods may be less efficient for such datatypes, since they may need to test. It is also something to keep in mind, whether during ufunc dispatching datatype instance specialization should be allowed (i.e. allow such a float32+float64 addition loop). But, possibly this can also be added later. The other thing is that it probably can always be possible to define a float32+float64 abstract DType.

Reviewer 7: It sounds then like you're proposing singleton instances for each of the builtin numeric dtypes? Without arguing about whether that makes sense or not yet, I think if that's really your intent, you should get the word `singleton` in a description somewhere further up.

Author: Right, almost singleton... We do have byte swapping (and unfortunately metadata). Eric, if you are unsure that is the correct path, should we chat about that briefly?

Author: Sorry, almost, since there is byte swapping, and unfortunately metadata.

Reviewer 7: > Eric, if you are unsure that is the correct path, should we chat about that briefly?

I need to set aside a reasonable block of time to finish my train of thought in that direction before such a chat would be useful, and large blocks of time are seemingly hard to come by.

Author: Yeah, that is always the problem (except for me, I got too much time to get lost in it). Maybe just for your consideration, why I am thinking along this line:

1. We currently have the methods (in `PyArray_ArrFuncs` defined for each type number basically) and which are housed very similarly to a class.
2. I am thinking that `isinstance(..., np.dtype[float64])` can be useful in general, so it might as well be its own class.
3. The more specific classes are, the easier it seems to quickly dispatch in the UFunc machinery. If float64 is its own class, a loop registered for `float64, float64 -> float64` does not need any additional checks to find the correct C-function (except possibly byte-swapping if you would want to optimize it/skip the casting).
4. I think a DType class is, especially for parametric dtypes, is the most reasonable basis for the first/main step in ufunc dispatching. It seems desirable that classes are thus as specific as useful in that step. It also makes incremental loop addition easier: you can define a ufunc for float64, and then someone else adds the float32 loop.

Of course there are some interesting thoughts, e.g. we could make most methods work more like `get_nonzerofunc(self)` style (why shouldn't nonzero optimize to skip the byte swap check).
But I suppose my angle is this: *It seems useful to me to make the classes as specific as conveniently possible. In other words fully specific aside from storage information (byte swap, string length) and parameters (string length, units).*
Basically, byte-order being considered inconvenient, otherwise it would be true singletons and the main reason to have instances would be for metadata, and because parametric dtypes need something like instances.

Reviewer 7: I started writing up some thoughts here: https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b.

I think I've convinced myself that dtype subclasses are the right way to go, and that metaclasses are not. There's one TODO left in there, which is to look at what `metaclass.__new__` does to store its state - this should give us some hints about where `metadtype.__new__` ought to store its state, and will hopefully put me on the right train of thought to think about singletons.

Author: Thanks, I will have to look through this (right now I am not even sure if you mean dtype as classes or a DTypeMeta to define the storage on the DType metaclass).

Reviewer 7: I'm afraid I wrote that after not having read this for a while, so it's based only off my memory of this NEP, and likely does not agree on terminology.

Author: Sure. I have tried to parse it. To be sure, you disagree with the metaclass thought to add something similar to `nb_numbers` slots (i.e. methods)? Or are you saying that things like `itemsize` and `flags` should not be be part of the a DType class (metaclass instance), which I fully agree with? I would argue that the "table of behaviours" *is* normally part of the type/class, while all the other things are *not*.

Author: I could see that some functions could be defined/specialized for the actual dtype instance (as used in this proposal, e.g. for byte-order or string length). I do think we have a few `staticmethod`s that we need, e.g. because we need to find the correct string length. Although of course there are other ways to approach that, for example using "flexible" instances as we are doing now (which cannot be attached to an array)...

Reviewer 7: I think what I wrote is more an argument against using the word "metaclass". dtypes are not types, so metadtypes are not metatypes. I've updated the doc with a table of comparisons between builtin scalar types and our array types. I don't think anything I've written disagrees with this NEP yet.

Author: Right, `dtypes` are not types, but instances of `DType` classes (types). For the user that is all they should care about.
For us, I do currently think we do need some low level C-storage on the DType classes as well as control over `__new__` so technically, I think the DType class has to be an instance of a `DTypeMeta` class, although if you have another idea I am happy to listen.

Its 3 levels deep, so confusing :(, but on the user side it should only be 2 levels `DType` classes (current type number) and `dtype` instances (current dtypes). Even implementors should not really have to care about it too much, they define a new `DType` class, much like a python class, but with slightly different API.

(Not sure this is actually helping, hmm)

Reviewer 7: > For us, I do currently think we do need some low level C-storage on the DType classes

In my tables, which "C decl" row are you thinking that state would need to live inside? I think that's covered without `DtypeMeta`.

You _would_ need `DtypeMeta` (or `IndexableTypeMeta`) to support `np.dtype[...]` as syntax, but that doesn't need to be in an NEP.

Author: Currently mainly:
 * Define some metadata about the type (kind, type_num, ...) (depending what `...` is, it may be here, or in the instance part).
 * fast-paths for "magic" methods (f)

Although the "magic" methods, include some like `common_dtype`, which I think are actually class methods (at least in my current view), because they are associated with a `type_num` more than a dtype. I.e. in my view it is odd that we have multiple instances with identical `f` and `type_num`, so those things should be stored on the class. (which effectively makes the class a MetaClass instance, although no python user has to ever think of it that way).

Reviewer 7: That wasn't a particularly clear question from my end, apologies. I was trying to point you to [the following section](https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b#comparing-scalar-types-and-array-types-in-more-depth), but maybe my question didn't make sense anyway.

> I.e. in my view it is odd that we have multiple instances with identical `f` and `type_num`, so those things should be stored on the class.

Note that this is how regular `types` behave today though. If you have:
```
class Foo:
    def __some_slots__(self):
        pass
class Bar(Foo): pass
class Baz(Foo): pass
```
Then at the C level, `((PyTypeObject*)Bar)->tp_some_slot == ((PyTypeObject*)Baz)->tp_some_slot == ((PyTypeObject*)Foo)->tp_some_slot`, thanks to the behavior of `Type_Ready`. Similarly, for python level method overriding / duplication, `type(Foo).__getattribute__` handles the work

So maybe the missing piece of the puzzle is a class heirarchy like:
```
|---- meta-dtypes -----------------|                 |- dtypes------------------------|
dtype <--issubclass-- integral_dtype <--isinstance-- int8 <--issubdtype-- int8_nonative
```
instead of what I think this NEP describes
```
|---- meta-dtypes --------------------------------------------|               |- dtypes----|
dtype <--issubclass-- integral_dtype <--issubclass-- int8_dtype <--instance-- int8
                                                     int8_dtype <--instance-- int8_nonnative
```
Where the distinction is simply _"Does it declare new hook or attribute names at the C level?"_
   * Yes: meta-dtype
   * No: dtype

Reviewer 7: I elaborated on that thought in https://gist.github.com/eric-wieser/49c55bcab744b0e782f6c2740603180b#what-this-could-mean-for-dtypes.

Author: @eric-wieser not sure I can follow, not on the top of my game today, but still. Yes, types use the same C-level slots, but *instances* do not typically. They are meant to be as minimal as possible (not that it matters for dtypes all that much).
One note that I do see `integral_dtype` as useful, but I do not really want to use inheritance for now, it just seems dangerous. E.g. a user integer should not inherit from it.

Note that in your class hierarchy, `int8` is an instance of `integral_dtype` and not `int8`. That moves the precision information into the instance. While possible I feel there is a disassociation with the current `dtypes.num` (which is vital for ufunc dispatching). There are two instances with the same `dtypes.num` (and `issubdtype`), but they have nothing directly in common.

I guess I see it as:

```python
class DType():  # capitalizing but same as np.dtype
   @classmethod
    def __common_dtype__(cls, other):
        raise NotImplementedError

    def __setitem_ptr__(self, ptr, value):
         raise NotImplementedError

class MyDType(DType):
    def __init__(self, itemsize=4):
        self.itemsize = itemsize

    @classmethod
    def __common_dtype__(cls, other):
        raise NotImplementedError

    def __setitem_ptr__(self, ptr, value):
         # does the current logic of ->f->setitem
```
Now in Python, those dunders will be stored in C-side slots on the class, I am suggestion to do the same thing. That requires a DTypeMeta to extend pythons `type` C-slots. While `itemsize` on the other hand is of course an instance variable stored on in the `DType` instance struct.

I am sure you realize all of this though. In my view `integral_dtype` is the same as a Python ABC, mainly something to define a hierarchy.

I guess my main question would be what about the current `dtype.num`? I would want that to be replaced by the class. `dtype.num` is IMO vital, because it is the basis for UFunc dispatching. Furthermore, I do not actually see much of an advantage of `int` being an instance of an `integral` class, it might work within NumPy, but if someone wants to add another integer I am not sure it helps much. And `int8` and `int64` are different enough to have different type numbers (and not all ufuncs may accept all of them).

My guess is what you are proposing is almost the same thing, but the split between what is an instance and what is a class is done at the `integer` level instead of the `int_of_given_precision` level. The main gain would be that you do not have these almost singleton instances. But I am not sure what the actual gain of that is, and I honestly do not quite see how we can allow someone to cramp their own ArbitraryPrecisionInteger into our `np.integer`, it would have to be a subclass? But it is not really a subclass by Liskov substitution principles.

Reviewer 7: > Yes, types use the same C-level slots, but _instances_ do not typically.

The parallel I'm trying to draw is:
1. _scalar instances_ &lrarr; _array instances_
2. _scalar types_ &lrarr; _array dtypes_
3. meta-types &lrarr; meta-dtypes.

With that parallel, it's 3 which declares the slot names, 2 which provides the values of the slots, and 1 which contains a pointer to the table of slots. This NEP seems to be leaning in the direction of

Author: I would prefer to write my proposal here as:

1. scalar instances ⇆ array instance + DType instance
2. scalar types ⇆ DType class (type)
3. Metatypes should not be a large priority for now, but `MetaType(type)` ⇆ `MetaDType(DTypeMeta(type))`

Let me try to discuss a few of the points, although it is getting difficult and I may have to take a step back and redefine the terminology here if we want to keep discussing (or make sure to better understand yours).

> This is a feature. isinstance(int8, integral_dtype) conveys the information "int8 carries extra C-level slots specific to integers", which could include
things like ->unsigned_int_type.

Right, however, like in Python were we rarely actually need MetaTypes, I do not think you actually normally need these additional slots on the C-level initially. I may want to enable this, but it seems not a priority to me (i.e. subclassing DTypeMeta). You can already tag on functions to types on the class (in python) well enough, as `int8.unsigned_integer_type()`. Yes, unless we allow subclassing DTypeMeta you cannot do so (efficiently) in C. I am not sure there is much need for it though, at least initially. First, many of these things can be just as well functions likely. Second, it should be possible to spell your example (maybe slightly less efficient) using `np.astype(np.int8, np.unsignedinteger)`.

So I agree with the sentiment that having only a single DTypeMeta (so to say) could be a limitation that should be waved, I am not quite convinced it has to be a priority right now, that integers have their own C-level slots.
To be clear, in my design, I *will* need these things to support things around value-based-casting. I am not opposed to making an `IntegerDTypeMeta` class, I just do not see it as important at this time, and would rather keep that API full private until someone comes along and actually needs it.

> I'd appreciate if you could elaborate on this point, being specific about the types you're referring to. I don't understand it yet.

Sorry, what I meant is that it seems to me for you `>int8`, `<int8`, *and* `int64` are all instances of the same `integer` class. Thus `>int8` and `<int8` must use their `dtype.num` slot to distinguish themselves from `int64`. Alternatively, the UFunc machinery, in a first step, would have to handle all integers identically. But UFuncs are currently designed around `dtype.num` slot. If `>int8` and `<int8` are instances of an `Int8` class (which *could* be an instance of an `IntegerDTypeMeta` if we have a good use for it in the future), then `>int8.type_num` can be replaced by `type(>int8)`, you do not need to store `dtype.num` on the dtype instance.

> It's probably worth comparing to how you'd build an ABC for a regular scalar type in python with C-level slots. My integral_dtype is the the new meta-dtype you
need to support the new slot definition, closer to ABCMeta.

Right (repeating the above), that is all good, it would be fine to have an `IntegerDTypeMeta` (in my terminology), I am just not convinced right now that it is necessary as a first step, especially as public API. I am not opposed though.

> Dispatching rules could use issubdtype(passed_type, dispatch_table_dtype), assuming we could make that fast enough.

Right, I personally feel using `type(dtype)` is a very elegant and easy to understand way though. I assume most speed will have to be fixed by caching anyway, but a `type(dtype)` lookup, is just an `id(type(dtype))` lookup, which should be easy enough to cache well.

> It's done at the "do I need new C attributes and function pointers" level.

Right, and I think a single `DTypeMeta` should be enough for that right now, if it is not, I am happy to start subclassing that `DTypeMeta`. Most Python scalars are also happy with a single metaclass (`type`).

> This sounds like me just choosing the wrong name, integer should have been fixed_integer

Fair enough, although you may want an `integral` as well, which I suppose also in your view could just be more like ABC though.

---

Let me be clear, I am not yet quite clear that we actually have different views here, except where to draw the lines.
I mainly think that a single (public) `DTypeMeta` is enough for now. I think users, and dtype implementors, should (normally) not have to worry about that `DTypeMeta` unless they are very advanced (very few Python programmers need MetaClasses), they just write a few dunders and let Python do the magic in `PyType_Ready`. Could more specialized `DTypeMeta` make sense? Yes, and we probably will need them, but I do not see them as a central concept.

Probably I am confused, but in your code snippet, I see `base_dtype == np.dtype` (sure it _could_ be something else, but I do not really see the need and am wary of implications). In your example `DType == np.dtype` seems wrong, since `DType` is a metaclass so `DType()` should be a class, but it currently is an instance (with an itemsize). Once you do that shift, and assume one `base_dtype == np.dtype` is fine, I think we may be saying very similar things?!

Reviewer 7: > in your code snippet, I see `base_dtype == np.dtype`

I don't see what you mean by that. Can you be more precise? (just edit your comment and I'll delete this comment)

Author: Hmm, I guess I mean that `base_dtype == np.dtype` is how I would read it, but I am not sure that you think that would be an OK thing to do (in general, I am fine if you say that it is not always what you would want to do).
Deleting both comments, but I am unsure that this actually clarifies things.

Author: @eric-wieser I am still struggling with what you write, in your gist:

> So we can support new dtypes defined using the class statement even if issubclass(np.dtype, type) is false.

Yes, IMO it should be `isinstance(np.dtype, type)`, now your `NotAType` seems a bit freaky? Since it has `not issubclass(type(NotAType), type)`. For one I do not see how it could possibly be initialized reasonably, since it will not have an `__init__()` and instead only has a `__call__()` since it is an object pretending to be a type. That is all fine, but it seems more confusing/limiting than helpful?

> dtype instantiation (not to be confused with invoking np.dtype itself) means just what it does today  [arr = uint8_non_native([1, 2, 3])]

This is not what it means today? That is the strange (I would prefer to deprecate) behaviour of *scalar* instanciation.

---

The only thing I currently managed to read out of your gist that you are thinking about putting *everything* on the same level as to remove the `class <-> instance` split I am proposing.
That is of course a possibility, but it seems weird for things like `metadata` (including "metadata" such as Unit information)? I do not think it is desirable to go through a subclassing (like) procedure/thinking to attach a different physical unit or string length. Also, I would like to be able to have some representation of `"S"` which is a string without a length. A string without a length is *not* a dtype in the sense that it cannot be attached to an array. For scalar-types, `np.integer` has no (direct) instances either. Representing it as the `DType` class (subclass of `np.dtype`) seems very elegant to me.
I do see that users are used to mainly work with dtype instances right now. And I do not expect that to change *most* datatypes. But I am unable to see the value in trying to flatten the distinction betwen DType classes defining the method/logic like a Python class and the dtype instance that is attached to an array. Yes, the (almost) singleton instances are a bit annoying, and the fact `np.dtype(np.int8)` can be mapped to `np.dtype("=int8")` while `np.dtype(np.string)` cannot be mapped (what is the strings length) *is* annoying. Since nobody will or should normally write `np.dtype[np.int8]()`...

Currently we have things like `.newbyteorder()` if you allow subclassing to add a "strange" byteorder, things will get tricky. In my model, every extension DType has to be its own class, there could sometimes be a shame (i.e. a UFunc would have to be told that it can use the `uint8` instead of the `uint8_strange_byteorder` loop. But I doubt it will be a big issue in practice and cleanly separates user datatypes and NumPy datatypes.

Reviewer 7: > Yes, IMO it should be `isinstance(np.dtype, type)`

Agreed, but the the two are orthogonal so I thought I'd be clear

> now your NotAType seems a bit freaky? Since it has not issubclass(type(NotAType), type). For one I do not see how it could possibly be initialized reasonably, since it will not have an __init__() and instead only has a __call__() since it is an object pretending to be a type.

It's not an object pretending to be a `type`, it's an object _being_ a `dtype`. `dtype`s have no `__init__`, because they're created via arrays, not directly. `NotAType` doesn't have a `__call__` either, but `type(NotAType)` does.

Conventionally, `class` definitions are a means of filling C slots in `PyTypeObject`. That example shows `class` definitions being used to fill slots in `PyArray_Descr`. What I'm trying to show is that it doesn't matter whether the slots being filled in belong to `type` or `dtype`, `class` is still a useful syntactic construct, and base classes still mean slot defaults.

> This is not what it means today? That is the strange (I would prefer to deprecate) behaviour of scalar instanciation.

Ah, that is a good point. It's still an example of how that could be done, but agreed it's not super necessary.

> The only thing I currently managed to read out of your gist that you are thinking about putting everything on the same level as to remove the class <-> instance split I am proposing.

Yes, I think that's the core of my proposal, which is motivated by the way that python handles scalar types. An interesting exercise would be to try and apply your ufunc-like dispatch system for a set of scalar only types implemented in pure python - would you need meta-types? Meta-metatypes? 

> That is of course a possibility, but it seems weird for things like metadata (including "metadata" such as Unit information)? 

Perhaps, perhaps not. Under that design, `issubdtype(np.int8.with_metadata(...), np.int8)` is true but not the reverse, which is perhaps tidier than today's behavior. If you wanted to create a regular `list` type with metadata, you'd subclass `list`, so it's sort of consistent

Regarding units, I think I should reread the NEP and try and translate the proposal here to how it would look in the gist.

Author: (Maybe replying too fast) OK, maybe we can focus on the "flattening" then (no instance <-> class split). One thing I am not 100% sure about is what `isinstance(np.dtype("int8"), np.dtype)` gives in your thoughts, I assume it is or can be made to be True (that is important for backcompat)?

> dtypes have no __init__, because they're created via arrays, not directly.

Fair point, but `arr.dtype` has no reason not to return the initialized object? It still seems strange to me to hide that initialization step from the user.
I do not disagree with the technical fact that it could be fine to not inherit from `type`, but doing it at least gives us useful `issinstance` and `issubclass` definitions?

> It's still an example of how that could be done, but agreed it's not super necessary. [About `int8([1, 2, 3, 4])` creating an array]

Yes, but that conflates the array object and the dtype, and I would like to keep them as separated as possible. That may end up being a fairly useless attempt, but it is a principle we should not give up quite yet?

> `issubdtype(np.int8.with_metadata(...), np.int8)` is true but not the reverse

Hmmm, I still feel `issubdtype(">i8", "<i8")` is a red herring in that it is a weird question to ask. In your thoughts it seems largely the same as `issubclass()`? But that cannot be commutative. In my proposal it would be deprecated and `isinstance(">i8", type(np.dtype("i8")))` would be preferred (np.dtype[np.int64]` and just `np.int64` are both acceptable alternative spellings for `type(np.dtype("i8"))`).

`issubdtype` supports the abstract scalar class hierarchy. For typical dtypes it is the same as comparing `type_num`, though, and I assume that `type_num` is a necessary concept – although maybe not in `issubdtype`.

This leads to a break in the hierarchy. `>int8` and `<int8` do not have an "order", they seemat same level but distinct e.g. from `<int64`. You could have an `all_int8s` class with  `<int8` and `>int8` as subclasses, but `<int8` and `>int8` are leaves of the hierarchy tree. And it seems to me that whether scalar or not, instances are the normal way to represent the leaves in such a tree.

I have to admit though, I am not sure what you are actually have in mind for `issubdtype(">i8", "<i8")` (error, True, or False).

> An interesting exercise would be to try and apply your ufunc-like dispatch system for a set of scalar only types implemented in pure python - would you need meta-types? Meta-metatypes?

I plan to propose multiple dispatching based on the `type(arr.dtype)`, with some limitations since I do not want to allow arbitrary subclassing for concrete DTypes (at the very least initially and w.r.t. inheritance). Just like a scalar multiple dispatching would use `type(scalar)`. So there is simply no difference IMO.

The problem is after that "dispatching" step you still could have different byte-orders, etc. so that casting may be necessary (ignoring fallback to loops that do not match exactly, by using abstract superclasses). There are no MetaClasses or MetaDTypes involved, but after dispatching the casting requires some of the `np.dtype` base-classes slots (dunders) – and yes, those require a MetaClass to look them up very quickly on the C-side.

> Regarding units, I think I should reread the NEP and try and translate the proposal here to how it would look in the gist.

I think even string may be enough. Because right now I do not see where the `string-of-given-length` factory would go naturally, in this proposal it is `np.dtype[np.string](length=10)` (where `np.dtype[np.string]` could have its own top-level name of course).
I am trying to abstract the "flexible" part away. "flexible dtypes" are classes, dtype instances cannot be "flexible".

Author: So I am trying to get back to this... and what the differences between my proposal here and your thoughts are.

The more I look at it, the less I see the relevant user differences to be honest, I have probably written similar stuff above.

You are proposing an "`issubdtype`" logic, when for me:
```
def issubdtype(dtype1, dtype2):
    assert isinstance(dtype2, np.dtype)
    isinstance(dtype1, type(dtype2))
```
(Although I have no desire to actually use subclasses for now, it keeps things simpler). We can probably agree that this hierarchy is necessary, e.g. for dispatching in ufuncs and casting. Whether `uint8_with_metadata is below uint8` matters little. 

My proposal has the downside that we have to use MetaClasses much more, and most relevant data is stored on those MetaClasses. That makes the dtype instances a bit thinner in theory, although that does not matter much. In practice, if you would want to have specialized fast-paths for some `>int8` functions, my approach is less nice. You could have an `>int8` subclass as well (making it basically true singleton instances), but that would require a fat type object.
To be honest: I do not think MetaClasses as an implementation detail matter, they are invisible for the user, fat objects do not matter for dtypes (unless you have hundrets of them).

So what are the actual differences?

1. Inheritance: `>int64` cannot actually inherit everything from from `<int64`. E.g. a comparison function is not byte-compatible. `int64` could inherit certain things from `integer` maybe. I doubt we want users to really use inheritance much in either case though.

2. `scalar types` ⇆ `dtypes or DTypes`. There are two sides to this "correspondence":
   a) In my proposal there is a direct mapping from a scalar type to the corresponding `DType`. This `DType` knows details such as storage and can be inspected by looking at its initializer. In particular the `Dtype` corresponds to an actual type and is not just a class (a set of values which can be represented).
   b) In your thoughts, you have a strong correspondence when it comes to methods. The `dtype` acts like a `type`, it is passed in the "instance", although here the instance is a pointer into the array.
  
  The main problem I see with that approach is how strings fit in. At least when it comes to strings, you fully lose the mapping property in a) in your proposal. Also you never have the type correspondence, you have the class-purpose correspondence in b).
  But I am not sure they fully correspond to a type in b) since the string length has to be passed in as well? I.e. you still need to pass in the `dtype` (as opposed to just its value), even though the method already lives on the dtype. That is not bad, but I think that the user-side correspondence in a) is more important? And passing in the `instance` as a `self` argument is a nice correspondence as well, and it is one that I have.
  In other words: I say that `scalar instance` ⇆ `array instance + DTypes instance`, which seems pretty clear about the fact that a method living on the `DType` needs both information from the array (the element) and the dtype?

So what are the actual downsides of the proposal in this NEP?

* I suppose I cannot optimize as easily for certain string lengths or specialize for byte-order on certain levels. This is semi intentionally, since we currently do not do it, but it is a possible downside.
* I have to wrangle with MetaClasses, I do not see this as a major issue though, mainly an annoyance.
* We have a lot of DTypes which have, for all practical purposes, singleton instances. And often we will have to allow you to pass in the DType and silently create the instance from it for you.
* I need abstract DTypes, and additional metaclasses to handle value based casting probably, that sucks... But, value based casting is a tricky beast, so, yes you could avoid having that require a MetaClass, but in the end the hacks will be similar, just housed in a slightly more lean object.

The upsides in my opinion are:

* We have a clear type hierarchy implemented the same way as the scalar type hierarchy and with classes. And the DType class/type corresponds to the actual scalar type (as in which values can be represented).
* The step of having `dtype` instances means parametric `DType`s have a clear API, `np.dtype[np.unicode]` can return something useful which can be *instantiated* to something useful.
* We can get rid of `issubdtype` or implement it directly based on the Python class hierarchy.

What Travis was thinking of, to get rid of the `DType` class would be to put everything on the scalar. Maybe what he was thinking of is that you would have a special method `np.int64.get_dtype(byteorder="=")`, which has its merit. In that view `type(np.int64.get_dtype(byteorder="="))` would be reasonable as your `integer` (although just `dtype` would be just as well within NumPy), and we might indeed store all dtype functionality on the instance instead of the class.
For Python builtin types we cannot do this of course, so this would mean we need some "dummy" scalars (and map to them).

But as much as I always like that, I somewhat fail to see that it will actually help much. We still need to define a bunch of structs (namely the current `->f`, etc.) which _effectively_ are much like defining a full scalar python class/type. A `np.int64.get_dtype()` method is neat, but I am not sure that returning an instance of a `np.dtype[np.int64]` is actually bad, at least it allows you to keep your scalars fully free of NumPy.

Author: OK, so to continue on that. If we say dtypes are simply a "generic" object describing the data, without additional bells and whistles, then I think the scalars have to take the brunt of it:

1. We will have to add at least some slots to the scalar:
    * Possibly just be the class method `__new_dtype__(*args, **kwargs)` and a method `__dtype__()`.
    * Casting needs a home somewhere, although it could be implemented more like ufuncs as multimethods if scalars were not MetaClasses.
2. This means that:
    * Scalars might need to be MetaClass instances, or `->f` slots need to be tacked on to every dtype instance (`->f` slots are much like a "class" IMO, but it may actually be neater/easier to avoid reference cycles between scalar types and DTypes).
    * Existing python scalars cannot add any additional methods. So we will need fake scalars, similar to `np.object_`, I think.
    * DType classes become an implementation detail. Subclassing `np.dtype` is only necessary for extension types that wish to add methods or store units, etc.. `np.isdtype` or similar exists, but you might as well work with `dtype.type` probably.
    * The scalar class hierarchy would be important if we want to use a hierarchy for dispatching! (which I currently see it being used for). Even if not, it would be the scalars providing the hierarchy in some form.
    * The DType classes would probably be considered to have less features, which may be good. The downside is if we have "fake" scalars like `np.object_`. If we have DType classes next to scalar classes, `np.dtype[scalar_type]` can always give you what you want (I admit doing the mapping is annoying). If someone wants to store `datetime.datetime` directly we need the same mapping somewhere, but there seems no "obvious" way to expose it to the user, which means there is no obvious way to do `datetime.datetime.get_dtype()` or similar. OTOH, I suppose you can still do a function approach. `np.dtype(datetime.datetime)` can still work, although it seems a bit tedious if the `datetime.datetime` DType were supposed to be parametric (where do you read up on the arguments/kwargs?).

 So... Maybe we can avoid metaclasses (as in special storage) by pushing things more into the dtype instances (more like it is now). A tricky part seems to me that we have to "identify" scalars that have DTypes, duck typing might work but we may need to use an ABC or even a MetaClass, which I am not sure I like to force up on scalars.

The biggest issue I still have is that it seems to me pushing things into the scalar seems tricky if the scalar is supposed to know nothing about NumPy. I.e. it would almost need to be a language feature in Python?
One of the larger annoyances I have with my metaclasses, is that the relation `Scalar Type<->DType type` may be a bit annoying. In some cases the scalar need not know anything about the `DType`, but in practice at least some of our scalars must. And I doubt that circular references even matter enough, the GC tracks tens of thousands of objects on a fresh IPython start...

Anyway, babbling by now :). I am not getting closer to being absolutely sure that one option is better than the other. And I am not sure that this actually has too much to do with your thoughts Eric.

In the end maybe my mistake is to think that normal Python scalars can/should have dtypes, and we should bake in that any scalar that has a "dtype" (can be stored directly, including discovery) to be derived from an `np.scalar`...

Reviewer 7: Hopefully I'll be able to come back to this conversation in the next week, just dropping in to assure you your thoughts are not going directly into the void.

Author: I find it hard to pin down the actual up and downsides. And of course your thoughts are probably in different direction anyway. Since I disliking the idea of a "hierarchy" outside of the class hierarchy. Of course there are arguments for using a "safe-casting" hierarchy, but I also see difficulties with that (although it sometimes generalizes a bit neater).

Up sides to a scalar-centric view:
  * Probably less MetaClassing
  * Less danger of circular referencing (although I doubt it matters)
  * No twin-implementation DType <-> Scalar type
  * No twin, possibly differing, hierarchies.
  * No twin implementation means no/less danger of random people implementing a DType for a random scalar.

Down sides:
  * Restricting inheritance/usage may be less desirable for scalars than for DType classes.
  * The scalar type must implement a certain interface (i.e. inheriting from `np.generic`, maybe even be a MetaClass instance). Or at least implement certain dunders. 
  * Thus: Difficult or impossible to implement DType for an arbitrary scalar type.
  * arrays rely on dispatching (unlike python scalars). The hierarchy is thus more important and allows for more confusion than typical Python scalars. Whatever hierarchy we use for ufunc dispatching may need to be cleaner than a typical Python scalar hierarchy.

User Impact:
  * DTypes such as `categorical` or `object` where the element is an arbitrary python object are more weird if the scalar type is central. But: We may actually say that `object` is the _only_ exception and a `categorical` result should be like an `enum.Enum` instance. A similar thing could be for an array of `np.ndarray`.
  * API Can either look more like:
    * `scalar_type.get_dtype(*args, **kwargs)`
    * `np.dtype[scalar_type](*args, **kwargs)`
    * If we allow dtypes without a clear scalar type, both have to use: `something(*args, **kwargs)`, in the scalar-centric case, I am not sure I have a good story, in the second you still need a `DType` class though. (The dtype type can be used also for the scalar-centric story, but I feel it has not quite fit in anywhere)

Implementation Impact:
* If the user has no scalar:
  * Scalar-centric: Create a new scalar with a NumPy provided API, or create a scalar and add a certain protocol on. They may need to create a `np.dtype` subclass as well.
  * DType-centric: Create a DType class and we have to help them to create a scalar as well.
* If the user already has a scalar:
  * Scalar-centric: The user has to extend the scalar and create a dtype subclass if necessary.
  * DType-centric: A DType is created, the scalar can remain unmodified (at least in principle)

---

But I guess, you may be coming from a bit different direction. And that is that the methods are actually more naturally defined on what we call `dtype` now, since it knows the exact data representation "type" of each element. And from that direction, I don't have much to add, except maybe
```
class Int8ScalarWithStorageDetails():
    byteorder = "="  # part of dtype
    byte_data 
```
also would store the byteorder on the instance.

Author: After a long time, in an attempt at resolving this discussion, or make my point of why this decision. I started to think about [here](https://hackmd.io/m6rhLhAqTmuhObW62N4sQA) (yes, its too long).

I will maintain what I think are my old points:

1. We need some representation of a "String" (i.e. `np.dtype("S")`).  There should be a clear place where value space and behaviour is *defined* (but not necessarily implemented).
   * `type(np.dtype("S5"))` can live up to that task
   * We have a clear relation that the value space and behaviour of one scalar-type is fully described by one DType 
   * In the model where the `dtype` instance is considered the implementer of behaviour, we need an
     abstract base class for each dtype: `np.dtype("S")` is *almost* that.
     We also need either that ABC or the scalar type to take the role of defining behaviour.
     (The scalar type is great, but I prefer it knows as little as possible about arrays and dtypes.)

2. Parametric `dtypes` are also normal python instances. There is some behaviour such as `np.promote_types()` and the NEP 42 `DType.__discover_descriptor__(python_object)` which has nothing to do with array values, and so it is clearly the `DType` class (meta-dtype) which should provide this behaviour.

At this point I said that I want something which "defines" (but not necessary implements) the value space and behaviour of a dtype.  And I claim, that the DType class (meta-dtype) is the natural choice, because we need it anyway and because otherwise the Python model would be to have an "abstract base dtype" (equivalent to our "flexible"  `np.dtype("S")` which technically shouldn't be attached to an array) instead.

So my perception is: We need a DType class and that DType class will *effectively* define/represent value space and behaviour.  Even in your model, where array element behaviour should be implemented on the `dtype` we have to deal with this DType class (it *might* be possible to make it more generic), but it must exist.

However: It does not need to *implement* the behaviour of the array elements.  We could look up behaviour just as well on the dtype directly.  Our current C code looks like that in places, since it uses:
```
PyArray_Descr *descr = PyArray_DESCR(arr);
descr->f->setitem(...)
```
Note that there is a *big* but here: `descr->f` is *always* shared for all dtypes with the same type number.  There is never any inheritance, so that the above code could just as well be:
```
PyArray_Descr *descr = PyArray_DESCR(arr);
descr->tp_type->setitem(...)
```
in all cases. And NEP 41 to, to some degree, requires this (see below to resolve this).

<details> <summary> In NEP 41, we consider the "instance" of a single array value to be `dtype+element`: </summary>

If we look inside an Python float, it has the structure:
```
float instance:
    type = float  # The instance stores its type
    value = 4.
```
The `dtype` itself is a Python type (spelling `type(np.dtype("float64"))` as `Float64`):
```
Float64 instance:
    type = Float64
    byte_order = "<"
```
While the individual array value can be thought of:
```
Float64 value:
    dtype = np.dtype("<float64")  # Stored on the array
    value = 4.
```
NEP 41, basically extends the value "instance"  rather than defining the type:
```
Float64 value:
    type = Float64       # stored in `arr.dtype`
    byte_order = "<"    # stored in `arr.dtype`
    value = 4.
```

</details>

---

The issue that I think we never talked about is the *why* you would want to store/implement behaviour on the `dtype` rather than its type.  And the main reason I can think of is to "specialize" functionality for speed.

My answer to the need of specialization is to think of methods similar to unbound methods in Python, i.e. this is what ufuncs do (or will do):

<details>

**Kernels (DType Methods) resolve the lack of specificity:**

A Python method is itself a class with a special `__get__` method. When the `.` operator fetches a method from a class, it binds it by calling `__get__`. We can write our own "method" class:
```
from functools import partial

def func(*args, **kwargs):
    print(args, kwargs)

class Method:
    _func = func  # function written by user

    def __get__(self, instance, owner=None):
        return partial(self._func, instance)

# Note that `__get__` is called and Method works:
class MyObject:
    meth = Method

bound_method = MyObject().meth  # calls __get__
bound_method()  # prints self
```

We can (and in casting to some extend do) translate this to
saying that a "kernel" (DType method), also can have
a `__get__` method:

```
class Kernel:
    _func1 = function_for_little_endian
    _func2 = function_for_big_endian
    
    def __get__(self, dtype):
        if dtype.byteorder == "<":
            return _func1
        else:
            return _func2
```
Making it possible to write functions specific to a a certain instance
(just as it would be possible in pure Python even *without* inheritance).
Although, the *organization* of these specific function occurs
one level higher.

This `__get__` method doesn't need to be exposed right away, but it is a clear
extension if "kernels" define at least almost all functionality of array values.
(Which is the clear goal with respect to ufuncs and casting.)

Once we expose the `__get__` function, the DType implementer could even create their own system of inheritance by looking up behaviour on the dtype instance itself (I guess similar to monkey patching a Python instance?).

</details>

Reviewer 7: I hope to find time to read this in the next month or so...

Instead it will be a subclass of ``np.dtype`` meaning that
``isinstance(np.dtype(np.float64), np.dtype)`` will remain true.
This will also add the ability to use ``isinstance(dtype, np.dtype[float64])``
thus removing the need to use ``dtype.kind``, ``dtype.char``, or ``dtype.type``

> to do this check.

Reviewer 7: In my opinion no one should have been doing the check this way anyway - the failsafe way to spell that has always been simply `dt == float64`, or `np.dtype(dt) == float64` if you don't know for sure whether you have a dtype or not.

Author: True, we do use the `kind` check for `np.issubdtype` checks in some places I think... But that is, to some degree, for later.
